module-services/service-cellular/ServiceCellular.cpp

```diff
@@ -923,10 +921,11 @@ auto ServiceCellular::receiveSMS(std::string messageNumber) -> bool
     }
 
     auto ucscSetRetries = 0;
-    while (ucscSetRetries < ucscSetMaxRetries) {
+    while (ucscSetRetries < at::AtCmdMaxRetries) {
         if (!channel->cmd(at::AT::SMS_UCSC2)) {
             ++ucscSetRetries;
             LOG_ERROR("Could not set UCS2 charset mode for TE. Retry %d", ucscSetRetries);
+            vTaskDelay(at::AtCmdRetryDelayMS);
```

Comment:
can you at least desribe why this is needed?

Reply:
For this command, no error was experienced so far. Delay is added just in case.

Reply:
to be completly honest - adding sleeps to the code "to be sure" doesn't sound like a good idea.

do you know how `channel->cmd` works?
it waits till there is set response from the modem, or error ...

Reply:
`channel->cmd()` executes a command and waits for the specific amount  timeout defined for each command in `module-cellular/at/src/ATFactory.cpp`. The delay this PR introduces is just a delay between retries of the command in case of command error.

Reply:
Actually, it waits for data - **or** - specific amount of time defined. The command timeout is meant to set a limit to wait time. We do not wait all that time each time we execute the command.
I am wondering - is the response immediate with error? is there a bus error returned or some other error? what's the exact cause which set `operator bool` invalid? How much time it takes to execute `channel->cmd()` execution?

Reply:
From implementation, the response is returned immediately in case of an error received from the `cellular`. If a timeout occurs, it's reported as well by the timeout result code.

Execution of `channle->cmd()` is timed in module-cellular/modem/ATCommon.cpp#L117, but it's not used for anything.

Reply:
I'm pretty sure that time is passed here:
`Result Channel::cmd(const std::string &cmd, std::chrono::milliseconds timeout, size_t rxCount)` isn't it? 

To be more specific:
1. what causes the at::Result::operator boo() const` returning false in this case?
2. how long doest it take to execute `(!channel->cmd(at::AT::SMS_UCSC2))` without sleep(50ms)? ( you can check it with 117 line, but you have to enable this first - see: `module-utils/log/api/log/debug.hpp` `DEBUG_SCOPED_TIMINGS`z

Reply:
RE 1: The error case covered by this PR is a deserialization error of malformed response frame:
```
104546 ms DEBUG [ServiceCellular] ATCommon.cpp:cmd:79: start of AT+QCMGR=0
104571 ms DEBUG [CellularMux] CellularMux.cpp:workerTaskFunction:554: DMA buffer full
104573 ms DEBUG [CellularMux] CellularMux.cpp:workerTaskFunction:554: DMA buffer full
{104573 ms DEBUG [CellularMux] CellularMuxFrame.h:deserialize:169: F9
[128 lines of byte dump ...]
104595 ms DEBUG [CellularMux] CellularMuxFrame.h:deserialize:169: F9 
}
104595 ms ERROR [CellularMux] CellularMuxFrame.h:deserialize:177: Received frame FCS [0xBD] != 0xCF error. Dropping.
104595 ms DEBUG [ServiceCellular] ATCommon.cpp:cmdLog:47: [AT]: >AT+QCMGR=0<, ><
104595 ms ERROR [ServiceCellular] ServiceCellular.cpp:receiveSMS:957: Could not read text message. Retry 1
```
RE 2: It takes 3 ms to execute `AT::SMS_UCSC2` command without sleeping for retry
```
71081 ms DEBUG [ServiceCellular] ATCommon.cpp:cmd:79: start of AT+CSCS="UCS2"
71084 ms DEBUG [ServiceCellular] ATCommon.cpp:cmdLog:47: [AT]: >AT+CSCS="UCS2"<, >OK<
71084 ms INFO  [ServiceCellular] ATCommon.cpp:cmdLog:52: [AT] > OK
71084 ms DEBUG [ServiceCellular] ScopedTime.hpp:~Scoped:35: Time to run at commandAT+CSCS="UCS2" time: 3
71084 ms DEBUG [ServiceCellular] ScopedTime.hpp:~Scoped:35: Time to run at commandAT+CSCS="UCS2" time: 3
```

Reply:
Yeah, but there is no information on how much it takes when there is an issue on DMA (from logs it seems like 50ms?) , and there is no information about the result in at::Resutl.

but from logs: **there is issue on DMA** and this is why I asked for result code returned in at::Result :(

Reply:
Applying some more thought through strategy to respond to errors here is out of scope for this PR and as @pholat mentioned, in fact the underlying problem seems to lie with DMA so we are simply applying a workaround here. Nevertheless, if we have no alternatives, lets apply those changes. Closing this conversation as discussed with @pholat.

Reply:
@pholat, and just 👍🏻, please.